File: two_sum/two_sum.go
package main

import (
	"fmt"
)

func twoSum(nums []int, target int) []int {
	tracker := map[int]int{}

	for x, val := range nums {
		counter := target - val

		if index, ok := tracker[counter]; ok {
		    return []int{index, x}
		}
			tracker[val] = x
		}
	return []int{-1, -1} // fallback
}

func main() {
	runTests()
	nums := []int{2, 7, 11, 15}
	target := 9
	result := twoSum(nums, target)
	fmt.Println(result) // Expected: [0 1]
}


func runTests() {
	tests := []struct {
		nums   []int
		target int
		expect []int
	}{
		{[]int{2, 7, 11, 15}, 9, []int{0, 1}},
		{[]int{3, 2, 4}, 6, []int{1, 2}},
		{[]int{1, 2, 3}, 7, []int{-1, -1}}, // no match
	}

	for _, test := range tests {
		result := twoSum(test.nums, test.target)
		fmt.Printf("nums: %v target: %d → result: %v (expect: %v)\n",
			test.nums, test.target, result, test.expect)
	}
}

File: two_sum/two_sum.py
#!/usr/bin/env python3


def two_sum(nums, target):
    tracker = {}
    for x in range(len(nums)):
        num = nums[x]
        counter = target - num
        if counter in tracker:
            return tracker[counter], x

        tracker[num] = x

    return -1


if __name__ == '__main__':
    nums = [2, 7, 11, 15]
    target = 9
    # Expected: [0, 1]
    print(two_sum(nums, target))
